wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsnext/SidecarBasedInternalEnvironmentProvider.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.workspace.server.wsnext;
+
+import java.util.Collection;
+import java.util.Map;
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.ValidationException;
+import org.eclipse.che.api.core.model.workspace.config.Environment;
+import org.eclipse.che.api.workspace.server.InternalEnvironmentProvider;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.environment.InternalEnvironment;
+import org.eclipse.che.api.workspace.server.spi.environment.InternalEnvironmentFactory;
+import org.eclipse.che.api.workspace.server.wsnext.model.ChePlugin;
+
+/**
+ * Creates {@link InternalEnvironment} from provided environment configuration, then converts it
+ * using {@link InternalEnvironmentConverter} and applies {@link WorkspaceNextApplier} to the
+ * resulting {@code InternalEnvironment} to add sidecar tooling
+ *
+ * @author Oleksandr Garagatyi
+ */
+public class SidecarBasedInternalEnvironmentProvider extends InternalEnvironmentProvider {
```

Comment:
SidecarBased - is this good name for this class. I mean not sure this class has true side-car specific. WDYT?

Reply:
This PR is POC that is supposed to be reviewed for the design not to stick with the implementation that won't be accepted by you or Gennady. So, I believe, that naming is not important at this stage.
Apart from that WS.NEXT is about sidecars, so I don't see why this name is bad. But you can always suggest another name.